[READY] perf improvements for strftime

smarie opened the pull request:
This PR is a new clean version of #46116

- [x] closes #44764
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
`period_format` now has a faster default formatter leveraging python string formatting

Commit:
class `_Period`: new method `fast_strftime`

Commit:
class `Timestamp`: new method `fast_strftime`

Commit:
New module in tslibs: `strftime.py`. New function in this module: `convert_strftime_format` (raises `UnsupportedStrFmtDirective`). This function converts a `strftime` date format string into a native python formatting string

Commit:
`format_array_from_datetime`: new boolean argument `fast_strftime` to use faster datetime formatting.

Commit:
datetimelike `strftime`: new boolean argument `fast_strftime` to use faster datetime formatting. `_format_native_types` modified with this new argument too. Subclasses modified to support it (`DatetimeArray`, `PeriodArray`, `TimedeltaArray`, `DatetimeIndex`)

Commit:
`DatetimeIndexOpsMixin.format` and `_format_with_header`: new boolean argument `fast_strftime` to use faster datetime formatting.

Commit:
`NDFrame.to_csv` and `DataFrameRenderer.to_csv` and `CSVFormatter.__init__`: new boolean argument `fast_strftime` to use faster datetime formatting.

Commit:
Added tests for the `to_csv` dataframe method to cover the new fast_strftime feature

Commit:
`TestCategoricalRepr`: added a test for dates without time, with timezone

Commit:
Fixed `test_nat` and `test_api` with the new symbols added

Commit:
New `test_strftime` module to cover the `strftime.py` module in tslib.

Commit:
`convert_strftime_format` argument `target` is now mandatory to avoid mistakes

Commit:
`convert_strftime_format`: Completed unsupported directives for datetimes

Commit:
Fixed bug in tslib `format_array_from_datetime`

Commit:
Fixed issue in `format_array_from_datetime` when tz was not None

Commit:
Added 2 todos

Commit:
`test_format`: Added various tests for the new feature

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

 Conflicts:
	pandas/_libs/tslibs/period.pyx
	pandas/io/formats/format.py
	pandas/tests/scalar/test_nat.py

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed Datetime64TZFormatter issue due to arg renaming in recent commits

Commit:
Added 2 asv benchs for strftime with iso8601 format, and a variant for tz-aware

Commit:
blackened, flake8, and removed asv main

Commit:
Minor improvement

Commit:
Added new ASVs for strftime

Commit:
Added asvs for period

WillAyd commented:
This PR is still pretty big. Any reason why you are introducing a new `fast_strftime` keyword instead of just trying to improve performance inplace? I think that would help to reduce the size, though still probably need to break up in smaller subsets. The bigger a PR is, the harder it is to review so ends up in a long review cycle

Commit:
Added asvs for datetime and period indexes .format

Commit:
`convert_strftime_format` is now part of the public API as it is required by `Period.fast_strftime` and `Timestamp.fast_strftime`

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

 Conflicts:
	pandas/tests/frame/methods/test_to_csv.py

Commit:
Updated whats new 1.5.4

smarie commented:
Thanks @WillAyd for the comment ! I filled the what's new doc to help with the review process.
I will also try to move part of the mods in a separate PR, I'll tell you when this is done

smarie commented:
@WillAyd  I created #51459 that is all the "independent" items that I could still find

smarie commented:
@WillAyd would you be able to review #51459 ? That would help remove some code from this PR

smarie commented:
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

Please leave this PR open, we are waiting for https://github.com/pandas-dev/pandas/pull/51459 to be merged to reduce the volume of changes, but otherwise this is ready

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

# Conflicts:
#	asv_bench/benchmarks/strftime.py
#	asv_bench/benchmarks/tslibs/period.py
#	doc/source/whatsnew/v1.5.4.rst
#	pandas/_libs/tslibs/period.pyx
#	pandas/_libs/tslibs/timestamps.pyi
#	pandas/core/arrays/period.py
#	pandas/tests/io/formats/test_format.py

Commit:
Fixed issues following the merge from latest main. Introduced new function `fast_period_format`

Commit:
Improved ASV bench slightly (added datetime index formatting tests)

Commit:
Merged whatsnew

Commit:
pre-commit and docstring checks

Commit:
blackened

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Added two asvs

Commit:
Fixed mypy error

Commit:
Fixed ASV bench

Commit:
Improved ASV benchs

Commit:
Fixed RST format in whatsnew

Commit:
Hopefully fixed the ASV bench for the case when the format is the default one but provided in an explicit way

Commit:
Fixed issue in period.pyx

Commit:
Made hooks happy

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

Commit:
Fixed variables used before initialization

Commit:
pre-commit hook upgrade

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed error in period.pyx

Commit:
Fixed issue: there was no need to encode to bytes to apply string formatting

Commit:
Fixed issue: there was no need to encode to bytes to apply string formatting

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new' into feature/44764_perf_issue_new

# Conflicts:
#	pandas/_libs/tslibs/period.pyx

Commit:
Revert "Fixed issue: there was no need to encode to bytes to apply string formatting"

This reverts commit d44a4c952dafa59762d915b6a2f464e9aa485f27.

Commit:
Fixed issue: there was no need to encode to bytes to apply string formatting

Commit:
Removed todo in timedeltas.pyx

smarie commented on pandas/_libs/tslibs/timedeltas.pyx:
```diff
@@ -1521,6 +1521,7 @@ cdef class _Timedelta(timedelta):
         comp_dict = self.components._asdict()
         comp_dict["sign"] = sign
 
+        # TODO make marginally faster using old-style python formatting ?
```
to remove

smarie commented:
This is now ready for review @MarcoGorelli or @mroeschke or @WillAyd .

The PR is still pretty big, but it is a single feature.
The only items I can think of removing to make it (marginally) smaller would be :

- these two test files : pandas/tests/arrays/categorical/test_repr.py and pandas/tests/io/formats/test_to_csv.py . Indeed the tests do not contain the fast_strftime parametrization.
- the ASVs

But it would not really reduce much the volume of the PR, and these items are also clearly related to the PR ; so not sure it is worth spending the time doing the splitting.

I propose to add the `fast_strftime` parameter everywhere mostly as a safety belt, 

- even if the tests are now covering all the aspects we can think of 
- even if there is an automatic fallback to the c-strftime when the formatting pattern can not be converted (`UnsupportedStrFmtDirective` caught in calls to `convert_strftime_format`).

Indeed, in spite of all the care taken in writing this PR the past 15 months (time flies ! original pr was #46116), if there is a problem, it will be much easier to indicate a workaround (`fast_strftime=False`) than a version downgrade (use `pandas<2.1`) in my opinion.

Commit:
Fixed asv bench

Commit:
Fixed ASV

Commit:
Fixed indentation issue

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Doc fix whatsnew entries sorted

Commit:
Improved speed of csv formatting of datetimeindex

Commit:
pre-commit fixes

MarcoGorelli commented:
thanks for updating

I'll take a look when I get a chance - perf improvements generally aren't highest-priority (compared with regressions / bugs / API changes), but this would be nice to have, the improvement is quite impressive. Just saying this to manage expectations, it may take some time to get this in
Generally speaking, do you know if the `strftime` performance is ever a bottleneck for anyone? It would be good improve it, but if I'm reading the results correctly, it doesn't look critical?

Not totally sure about having a separate option though, would rather not introduce complexity. Is there any case when you can envisage this failing?

smarie commented:
> Generally speaking, do you know if the `strftime` performance is ever a bottleneck for anyone? It would be good improve it, but if I'm reading the results correctly, it doesn't look critical?

Well, it is actually quite critical for industrial applications. Actually, the very reason why I opened this thread of PRs more than one year ago was that our cloud applications in production rely on pydantic + pandas to expose dataframes as payloads in web services, and we identified that our main bottleneck in perfs was indeed pandas strftime.

MarcoGorelli commented:
makes sense, thanks for explaining!

smarie commented:
> Is there any case when you can envisage this failing?

No, I think that I found most edge cases already. The mechanism is based on two items:

 - the `convert_strftime_format` template converter, that moslty contains string "is in" and "replace" statements https://github.com/pandas-dev/pandas/pull/51298/files#diff-fd90258af6c8946c90f0af064c63f1a5941102ba948e798c2c5b248308197219R133
 - the python %-formatting mechanisms with keywords https://github.com/pandas-dev/pandas/pull/51298/files#diff-81022a29ff44a1036a4aca0409893e34a78526bf7b5a016171fa9f4bd743bc83R1484 https://github.com/pandas-dev/pandas/pull/51298/files#diff-6e86989165ac69c298669127b6fefcc0807ac6d601c61677a9348b2ec46ef433R231
 
Both seem simple enough to not hide unexpected behaviours. However maybe on some platforms (Pypy, embedded...) some users will want to use the c-strftime ? For some reason that I don't know ?

smarie committed:
Removed invalid comment

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed typo

smarie commented:
Hello @MarcoGorelli , did you have a chance to start reviewing this ?

Concerning the removal of the `fast_strftime` argument, maybe we should at least keep it for the few "core" functions (the low level ones) so as to ease maintenance tasks. Maybe when reading the PR you will have a better idea of where this can remain useful ?

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed whatsnew

Commit:
Fixed changelog indentation error: now a one-liner

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed mypy error

Commit:
Fixed mypy error 2

Commit:
Fixed whatsnew

Commit:
Fixed whatsnew ?

Commit:
Fixed whatsnew ?

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

smarie commented:
@MarcoGorelli I finally managed to have all green lights from CI :)

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

smarie commented:
> If this is too much to review, maybe I can split this PR in three:
> 
>     * first only for timestamps (not periods).
> 
>     * then CSV formatting
> 
>     * then period and periodarrays
> 
> Would that help you ?

Gentle ping - @MarcoGorelli , just to get you opinion

MarcoGorelli commented:
hey - sorry for the delay, I was away the past two weeks

I have some priorities at the moment I'm afraid, and very limited maintenance time available - but yes, any effort to split it up would certainly help, thanks

smarie commented:
Thanks @MarcoGorelli for taking the time to answer. Ok, I'll try to split this in 3 in the upcoming weeks.
Also if other reviewers already familiar with the previous similar PRs can help, it can maybe be a solution to accelerate review time?
 (@mroeschke , @jreback or @jbrockmendel  )
Thanks again, we'll see how this goes

MarcoGorelli commented:
sure - will try to get this into the 2.x series anyway, if not 2.1 it'll be 2.2

MarcoGorelli commented:
Just to disclose priorities - a couple of high-priority items I think really need addressing for the 2.x releases:
- reviewing #52064
- addressing #34916

If you felt like helping to address either, then the sooner we can get those in, the sooner we can move forwards with this one

Thanks for your patience and understanding generally, interactions with you have been very positive

smarie commented:
Thanks @MarcoGorelli for the positive feedback !
If I can find some time one of these weeks I'll have a look, otherwise I'll wait until this is completed

MarcoGorelli reviewed (changes requested):
sorry, this doesn't look right
```python
(Pdb) p idx
DatetimeIndex(['2000-01-01'], dtype='datetime64[ns]', freq=None)
(Pdb) p idx.strftime('%-d')
Index(['%-d'], dtype='object')
```

smarie commented:
> ```python
> idx.strftime('%-d')
> ```

Thanks @MarcoGorelli ! Indeed this is not recognized as a valid pattern. I do not know this pattern, it is not indicated in [python docs](https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior) but still appears to be recognized by the C strftime. Still, I cannot find any explanation about it in the doc: https://man7.org/linux/man-pages/man3/strftime.3.html

Is this a nominal behaviour or a side-effect ?

If this is desirable (and I guess so for retrocompatibility), but we do not manage to find an exhaustive documentation of all this desirable behaviours, then I supposed that the safest fix is to replace [these lines](https://github.com/pandas-dev/pandas/pull/51298/files#diff-fd90258af6c8946c90f0af064c63f1a5941102ba948e798c2c5b248308197219R270-R271) by "if there is a remaining `%`, raise an error"

MarcoGorelli commented:
ah maybe it's OK, thanks - I was looking at https://strftime.org/, but it says "platform-specific". still, would be good to preserve behaviour there if possible

MarcoGorelli commented:
> then I suppose that the safest fix is to replace [these lines](https://github.com/pandas-dev/pandas/pull/51298/files#diff-fd90258af6c8946c90f0af064c63f1a5941102ba948e798c2c5b248308197219R270-R271) by "if there is a remaining %, raise an UnsupportedStrFmtDirective error". This will trigger a fallback to the legacy strftime usage.

sounds fine

can we remove the `fast_strftime` option if we want to get this in for pandas 2.x please? if we're confident in it, let's just use it

can we also add a hypothesis test for this please? something like
```python
import hypothesis.strategies as st
from hypothesis import given
import datetime as dt
@given(
    datetimes=st.lists(st.datetimes(min_value=dt.datetime(1700, 1, 1), max_value=dt.datetime(2200, 1, 1))),
    fmt=st.sets(
    st.sampled_from(
        [
        '%a',
        '%A',
        '%w',
        '%d',
        '%-d',
        '%b',
        '%B',
        '%m',
        '%-m',
        '%y',
        '%Y',
        '%H',
        '%-H',
        '%I',
        '%-I',
        '%p',
        '%M',
        '%-M',
        '%S',
        '%-S',
        '%f',
        '%z',
        '%Z',
        '%j',
        '%-j',
        '%U',
        '%-U',
        '%W',
        '%-W',
        '%c',
        '%x',
        '%X',
        '%%',
        ]
    ), min_size=1)
)
def test_me(datetimes, fmt):
    fmt = ''.join(fmt)
    idx = pd.DatetimeIndex(datetimes)
    result = idx.strftime(fmt)
    expected = pd.Index([i.strftime(fmt) for i in datetimes])
    pd.testing.assert_index_equal(result, expected)
```

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

smarie commented:
Sorry @mroeschke and @MarcoGorelli I have had much trouble finding time to work on this lately, but I still aim at completing it in the upcoming weeks.

MarcoGorelli commented:
awesome thanks - feel free to ping us when we should reopen (or just open a new PR, up to you)

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

# Conflicts:
#	pandas/_libs/tslibs/period.pyi
#	pandas/_libs/tslibs/period.pyx
#	pandas/core/arrays/datetimes.py
#	pandas/core/arrays/timedeltas.py
#	pandas/core/indexes/datetimelike.py
#	pandas/io/formats/csvs.py
#	pandas/io/formats/format.py
#	pandas/tests/io/formats/test_format.py

Commit:
Moved whatsnew items

Commit:
Re applied mods to `TestPeriodIndexFormat`, since it had moved somewhere else

Commit:
Re applied mods to `TestDatetimeIndexFormat`, since it had moved somewhere else and was renamed since #55603

Commit:
Re applied mods from `TestFastStrfTimeScalars` in the right places

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new' into feature/44764_perf_issue_new

smarie commented:
@mroeschke @MarcoGorelli : please reopen, I have updated with the latest main branch. (https://github.com/pandas-dev/pandas/pull/55603 was a bit painful , note for myself. But I think I managed to find the "new" right destination for all tests)

Once reopened, I will apply the two changes we discussed above (removing faststrftime arg everywhere and defaulting to legacy method in presence of an unresolved %)

Thanks!

Commit:
Ruff: Fixed invalid character in comment

Commit:
black+isort+fixed tests

Commit:
Implemented conservative fallback as suggested per code review

Commit:
black

Commit:
Implemented hypothesis tests as suggested per code review

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Removed the `fast_strftime` argument everywhere as suggested per code review

Commit:
Fixed docstring

Commit:
Trying to have meson understand that there is a py file in the tslib

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed ruff error

Commit:
Fixed tests

Commit:
Fixed test

Commit:
Fixed test

Commit:
Fixed test

Commit:
Fixed test for musl linux

Commit:
Fixed mypy errors

Commit:
Fixed mypy error

smarie commented:
@mroeschke @MarcoGorelli : now ready for review.

All tests pass, the remaining CI failures are not relevant to this PR it seems, but to a warning in numpy related to encoding when scipy `sem` is used

Commit:
Fixed mypy error

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

smarie commented:
@mroeschke @MarcoGorelli : would you have bandwidth to review this PR before end of year ? We could make it in 2023 :)

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

MarcoGorelli commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -1457,6 +1457,37 @@ class Timestamp(_Timestamp):
         tz = maybe_get_tz(tz)
         return cls(datetime.fromtimestamp(ts, tz))
 
+    def fast_strftime(self, fmt_str: str, loc_s: object) -> str:
+        """A faster alternative to `strftime` using string formatting.
```
any reason why there's still a new function?

In https://github.com/pandas-dev/pandas/pull/51298#issuecomment-1655952627 I'd suggest to add a hypothesis test - if we're confident in it, let's just use it and speed up where possible

smarie replied:
This instance-related one is a low-level utility function for internal calls. Its inputs are the outputs of the string template converter, so it is only for debugging purposes. If you prefer I can add an underscore prefix for this one.

> I'd suggest to add a hypothesis test - if we're confident in it, let's just use it and speed up where possible

Is the current hypothesis test not enough ? I had inserted it in response to that old comment of yours. https://github.com/pandas-dev/pandas/pull/51298/files#diff-119789a2741b06c31031dfb70a5cbd9aab84d854dec9b649e2fd67c76ff00531R970

smarie commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -97,6 +97,17 @@ This future dtype inference logic can be enabled with:
+
+Strftime performance improvements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+- New :func:`convert_strftime_format` to convert a strftime formatting template into a python string formatting template.
+- New :meth:`Period.fast_strftime` and :meth:`Timestamp.fast_strftime` leveraging templates created with :func:`convert_strftime_format`
```
@MarcoGorelli this relates to your comment. As you can see even if the `fast_strftime` argument is now hidden and automatically done whenever possible, the `convert_strftime_format` tool is still public (for maintainance/debugging/development purposes) and therefore `Period.fast_strftime` and :meth:`Timestamp.fast_strftime` can stay public for advanced users. This does not change the fact that these are used in a fully automated way whenever possible.

MarcoGorelli replied:
thanks - OK should have capacity to get to this next week, let's see if we can get this in

smarie replied:
That would be awesome. For the instance-related `Period.fast_strftime` and `Timestamp.fast_strftime` we have three options:

 1. current version "as is"
 2. adding an underscore prefix as mentioned above
 3. dropping these methods and inlining them instead. Not sure that it would bring something, but it would certainly imply to modify some of my tests.

smarie commented on pandas/io/formats/format.py:
```diff
@@ -1647,10 +1657,26 @@ def get_format_datetime64(
         )
     else:
+        # Relies on datetime.str, which is fast already
+        # TODO why is date_format not used in this case ? This seems like a bug?
```
The only place where this function is used with a non-None date_format seems to be within the `_Datetime64TZFormatter` class. 

I do not clearly understand from the code base the purpose of this class, as it is not used in the CSV formatter for example. Maybe it is deprecated ? Anyway, the type hint for the `date_format` attribute is `None` suprisingly, so it explains why not using it is not causing much trouble : https://github.com/pandas-dev/pandas/blob/70fc17444be82a8e07bfc1b69088379b79a3653b/pandas/io/formats/format.py#L1493

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

MarcoGorelli commented:
didn't manage to finish this, but I've put it in my tasks for next week

quick comment - I wouldn't introduce `pd.convert_strftime_format`, I don't think it should be a top-level function. maybe in `pandas.tseries.api` (I think that's it)?

Commit:
Removed `convert_strftime_format` from top-level API and moved it to `pd.tseries.api`.

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new' into feature/44764_perf_issue_new

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

Commit:
isort

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new' into feature/44764_perf_issue_new

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

smarie commented:
Indeed. I made the change, it is now only in `pd.tseries.api.convert_strftime_format`

MarcoGorelli commented:
thanks

I'm going to have to focus on getting https://github.com/pandas-dev/pandas/issues/56503 in for 2.2.1 (hopefully), so pushing this off my stack temporarily, but let's get it in for 3.0

smarie commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -2504,6 +2581,41 @@ cdef class _Period(PeriodMixin):
         object_state = None, self.freq, self.ordinal
         return (Period, object_state)
 
+    def fast_strftime(self, fmt_str: str, loc_s: object) -> str:
+        """A faster alternative to `strftime` using string formatting.
+
+        `fmt_str` and `loc_s` should be created using
+        `convert_strftime_format(fmt, target="period")`.
+
+        See also `self.strftime`, that relies on `period_format`.
+
+        Examples
+        --------
+
+        >>> from pandas._libs.tslibs import convert_strftime_format
+        >>> a = Period(freq='Q-JUL', year=2006, quarter=1)
+        >>> a.strftime('%F-Q%q')
+        '2006-Q1'
+        >>> fast_fmt, loc_s = convert_strftime_format('%F-Q%q', target="period")
+        >>> a.fast_strftime(fast_fmt, loc_s)
+        '2006-Q1'
```
Indeed, fast strftime is really only making sense for arrays.

I introduced the instance-related methods mostly for convenience, for debug. I can either remove them entirely or prefix them with `_`.

I am not very fond of merging them with the public `strftime` method as I don't see "normal" users using the fast method, this is only a tool to debug and possibly improve the feature later.

smarie commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -97,6 +97,17 @@ This future dtype inference logic can be enabled with:
+Strftime performance improvements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+- New :func:`pd.tseries.api.convert_strftime_format` to convert a strftime formatting template into a python string formatting template.
+- New :meth:`Period.fast_strftime` and :meth:`Timestamp.fast_strftime` leveraging templates created with :func:`convert_strftime_format`
```
Agreed. (see my previous comment too)

what about making them private ? (and removing from changelog then)

MarcoGorelli replied:
yup that sounds fine thanks

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

# Conflicts:
#	pandas/tests/arrays/test_datetimelike.py
#	pandas/tests/indexes/datetimes/test_formats.py
#	pandas/tests/indexes/period/test_formats.py
#	pandas/tests/scalar/period/test_period.py

Commit:
As per code review: instance level `fast_strftime` are now private `_fast_strftime`

Commit:
Added maintenance comment

Commit:
black+isort

Commit:
Moved changelog to 3.0.0 and improved it slightly

Commit:
Merge branch 'feature/44764_perf_issue_new' of https://github.com/smarie/pandas into feature/44764_perf_issue_new

Commit:
Fixed test: fast_strftime not part of api anymore (private)

smarie commented:
I made the two methods private. Many errors still happen in CI, not sure why they pop up now. I'll have a look next week, don't hesitate to let me know if some of this known and not related to the PR (such as the numpy dev one)
Thanks!

smarie commented:
The failing jobs does not seem to be related with the PR : 

 - numpy dev
 - `TestTimestampRendering.test_repr` test failing on windows and linux musl only when timezone is `dateutil/US/Pacific`
 - `test_append_with_timezones` failing on windows (not musl) because `US/Eastern` seems to be replaced with `America/New_York`.

These issues are probably related with a new version of pytz, not with the PR ? On my windows machine none of the above tests fails, I have `pytz==2023.3.post1    (pyhd8ed1ab_0)`

Let me know

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

# Conflicts:
#	pandas/tests/arrays/test_datetimelike.py

Commit:
formatting

Commit:
Fixed ASV benchmark: format method disappeared from DateTimeIndex

Commit:
Improved what's new

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

smarie commented:
All should be ok now including ASV. Waiting for the build to end to get confirmation :)

Commit:
Fixed ASV error

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

MarcoGorelli reviewed (changes requested):
thanks for updating

I think it's not correct with negative dates:
```python

In [21]: pd.DatetimeIndex(np.array(['-0020-01-01', '2020-01-02'], 'datetime64[s]')).strftime('%y')
Out[21]: Index(['80', '20'], dtype='object')
```

smarie commented:
> thanks for updating
> 
> I think it's not correct with negative dates:
> 
> ```python
> In [21]: pd.DatetimeIndex(np.array(['-0020-01-01', '2020-01-02'], 'datetime64[s]')).strftime('%y')
> Out[21]: Index(['80', '20'], dtype='object')
> ```

Indeed. But `pandas` is not consistent with negative dates currently. Indeed instance strftime raises an exception:

```python
pd.DatetimeIndex(np.array(['-0020-01-01', '2020-01-02'], 'datetime64[s]'))[0].strftime("%y")
```

raises

```python
NotImplementedError: strftime not yet supported on Timestamps which are outside the range of Python's standard library. For now, please call the components you need (such as `.year` and `.month`) and construct your string from there.
```

Note that the same error is raised by the equivalent more straightforward `Timestamp("-0020-01-01").strftime("%y")`

MarcoGorelli commented:
Raising "not yet supported" is fine, we really need to avoid silently returning wrong results

smarie commented on pandas/_libs/tslibs/meson.build:
```diff
@@ -51,6 +51,7 @@ sources_to_install = [
     'offsets.pyi',
     'parsing.pyi',
     'period.pyi',
+    'strftime.py',
```
From what I understand, Meson seems to require the list of all py and pyi modules to include (=the non-pure-C symbols). Without this line the symbols defined in strftime.py are not included in the tslibs sub-package if I remember correctly.

smarie commented:
> Raising "not yet supported" is fine, we really need to avoid silently returning wrong results

It would slow down the performance of strftime on arrays if we were to raise an error. So I suggest to rather fix it, the fix is fairly simple. 

However, this will create a difference between the more permissive array operation strftime, and the legacy less permissive instance-level strftime.

For this reason, I suggest to add an explicit argument `fast: bool=True` to the instance-level method. That way, by default the instance and array ops will have the same behaviour. Yet, on developer's explicit choice, turning fast=False on the instance method would use the legacy slow method.

What do you think ?

Commit:
Code review: renamed `loc_s` into `locale_dt_strings`

Commit:
Code review: moved whats new entry to perf

Commit:
Code review: removed commented out code

Commit:
Code review: added support for negative and small years and added corresponding tests.

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new' into feature/44764_perf_issue_new

# Conflicts:
#	doc/source/whatsnew/v3.0.0.rst

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Improved test for small years

Commit:
Fixed failing test

Commit:
Fixed failing test

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

Commit:
Fixed test on linux

Commit:
Trying to debug datetime.strftime on linux

Commit:
Fixed tests on linux and windows

Commit:
Removed useless comment

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

smarie commented on pandas/tests/scalar/timestamp/test_formats.py:
```diff
@@ -199,3 +217,84 @@ def test_repr_matches_pydatetime_tz_dateutil(self):
         dt_datetime_us = datetime(2013, 1, 2, 12, 1, 3, 45, tzinfo=utc)
         assert str(dt_datetime_us) == str(Timestamp(dt_datetime_us))
+
+    @pytest.mark.xfail(
+        is_platform_linux(), reason="strftime on linux does not zero-pad %Y"
+    )
```
I think I found a way. I'll if/else in the test

Commit:
Fixed tests on linux musl

Commit:
Attempt to fix on MUSL

Commit:
Fixed test for linux musl

smarie commented:
I thought of another side effect of current proposal, hopefully minor but I let you decide.

If the user uses a strftime template that can not be converted to the fast python template, for example `%Y %a`, then `DateTimeIndex.strftime` will silently fallback to current behaviour, that is, OS-dependent C strftime.

Therefore a user could experience this situation : 

 - "It was working fine with negative dates but when I modified the template from '%Y' to '%Y %a' it stopped supporting negative dates".
 - or "on my linux machine, %Y was providing a non-zero-padded representation of the year for year 123, and when I changed the template from '%Y' to '%Y %a' it now provides a zero-padded representation"

Both of these situations can be solved easily by reintroducing a parameter on all array-acting `strftime`, controlling which engine is used. For example `DateTimeIndex.strftime(format: str, engine: str = "auto" (default), "py_str_template", "c_strftime")`. The default behaviour would be the current one (silent fallback from `'py_str_template'` to `'c_strftime'` if `format` cannot be transformed into a fast string template).

Let me know how you want to proceed.

smarie commented:
I had a few thoughts about all of this during week end and decided to describe the two related issues of consistency and error management in separate tickets : 

 - consistency : https://github.com/pandas-dev/pandas/issues/58179
 - error management : #58178

That way, we can potentially try to not solve all issues at once, merge the current PR purely as a "performance" PR, and discuss the longer-term vision in dedicated PR(s) tackling the two above tickets.

MarcoGorelli commented on pandas/_libs/tslibs/strftime.py:
```diff
@@ -0,0 +1,278 @@
+
+    # Mapping between strftime and string formatting, according to both styles
+    if new_style_fmt:
+        esc = "/_+\\"
+
+        # Escape the %% before searching for directives, same as strftime
+        strftime_fmt = strftime_fmt.replace("%%", esc)
+
+        esc_l = "+^_\\"
+        esc_r = "/_^+"
```
this looks a bit mysterious to me, mind explaining why you've constructed `esc`, `esc_l` and `esc_r` like this?

smarie replied:
haha good catch. Indeed this is ugly. Purely random strings here. Long enough so that users will not hit them by chance. 

A more robust way would probably be to check that the above do not exist in the input sequence, and if they do, to add as many trailing '+' (or any other char) as needed so that it is not present in the input string. We could implement it in a `find_escape_patterns` internal function. 
If you agree with this strategy I can change it like this.

smarie replied:
fixed

MarcoGorelli reviewed (commented):
Nice one! I have a question about `esc`, but I think this looks good

Agree with your suggestion about adding `engine`, and agree that it can be done separately

Thanks for having stuck with this for so long - the perf improvements are really noticeable so it's worth doing this!

@WillAyd do you have any comments?

WillAyd commented:
Sorry if I have missed the decision in comments but from an API perspective I really dislike `fast_strfrtime` - why are we not just improving performance in place?

smarie commented:
@WillAyd there is no `fast_strftime` public API in the current proposal. There are new private methods on `Timestamp` and `Period` instances, and there is a new parameter in the low-level C array routines of tslib for period and timestamp arrays.

The reason we cannot completely hide these is due to the fact that the legacy engine is OS-dependent and has a larger compatibility scope, while the new faster engine is non-OS-dependent and has a more restricted compatibility scope. This is explained in details in the past discussions above, and summarized in the "why is it so" section of https://github.com/pandas-dev/pandas/issues/58179

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

# Conflicts:
#	doc/source/whatsnew/v3.0.0.rst
#	pandas/tests/scalar/timestamp/test_formats.py

Commit:
Simplified `convert_strftime_format` as per code review

Commit:
Renamed `fast_strftime` with `strftime_pystr` as per code review

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new

Commit:
Fixed test on WASM/Pyodide

smarie commented:
Sorry for the delay in answering those last comments @MarcoGorelli and @WillAyd.
I simplified the escape characters creation @MarcoGorelli , and removed the name "fast_strftime" everywhere, replacing it with "strftime_pystr" @WillAyd .
Hopefully this will now be ok for both of you.

Commit:
Renamed internal argument ``strftime_pystr`` into ``_use_pystr_engine`` to be more explicit.

Commit:
Fixed test on pyodide

Commit:
Fixed cython-lint error

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new

MarcoGorelli commented:
thanks for updating!

> from an API perspective I really dislike fast_strfrtime

agree, but it's not part of the public-facing API - are you OK with having an argument to trigger the fast path just in internal functions?

WillAyd commented:
Haven't done a full deep dive on the latest changes yet but at a high level seems reasonable, though maybe would opt for just `engine="python" | "c"` since that is a precedent we have in other methods (granted, C is normally the faster option)

Would that alleviate your concern @MarcoGorelli

MarcoGorelli commented:
sure, but there are cases when the `'c'` engine isn't supported

so, then, pandas could either:
- fallback to `engine='python'`
- raise, telling the user to pass `engine='python'`
- use `engine='python'` as default

smarie commented:
Please see my proposal in https://github.com/pandas-dev/pandas/issues/58179 under "feature description". I mention 3 parameter values (equivalent of 'python', 'c', and 'auto'. The 3d one is the missing piece in your discussion here I guess

MarcoGorelli commented:
hi @WillAyd - I think Will might be asking why `_use_pystr_engine` is even an argument of `format_array_from_datetime`, if it's always set to `True` when calling it. I think it could just be removed in this PR, and `format_array_from_datetime` determines whether to use the fastpath or not?

smarie commented:
@WillAyd do you confirm that removing the `_use_pystr_engine` parameter from `format_array_from_datetime` would do the trick for you ? I would prefer to be sure that this is the last and only change remaining for this PR. Once merged, I will switch to the "engine" parameter mentioned in https://github.com/pandas-dev/pandas/issues/58179

If you would like more changes to be done in this PR, do not hesitate to let me know now, so that I can tackle all changes in the same update. Indeed multiple very small changes induce a large overhead on my side unfortunately (dev env, pre-commit checks, ci/cd, tidying everything, etc.). 

Thanks in advance for your understanding and help to get this feature out !

smarie commented:
@WillAyd @MarcoGorelli , is there any way we could wrap up this ? We are very, very close to a success here - let me know how you want to proceed

smarie commented:
Thanks @WillAyd for this feedback ! 

> The fact that the "python" implemenation is faster than the "c" implementation is backwards from the rest of our API

Indeed, engine names proposals `py_str_template` (faster) vs. `c_strftime` (slower) is counter intuitive. I think that this API consistency can be solved simply by using better names:
- for the faster version leveraging stdlib string templating: `cpy_str_template` (referencing to the fact that the python stdlib is actually implemented in C and optimized for speed), `stdlib_str_template` or `stdlib_str_format` or `str_format` or `str.format`. Note that any remaining ambiguity can be easily solved by a clear docstring
- for the slower current version leveraging OS-specific implementation of strftime, maybe using `os_strftime` instead of `c_strftime`, will be easier to not relate to any speed expactation. Here again a clear docstring will do the trick to solve any remaining ambiguity about why this is not the default engine.

> The other consideration point is pyarrow - is this faster than that? From the descriptions provided, it seems like we are still heavily using Python strings, which I would think is a bottleneck. I expect it would be much simpler just to dispatch to pyarrow.compute and circumvent this altogether?

`pyarrow` is definitely an engine worth looking at, that can be added to the list of engines available in the future, and can even become a default one if we can validate that the implementation suits all of pandas community expectations. I see that this function is available : https://arrow.apache.org/docs/python/generated/pyarrow.compute.strftime.html . Still I don't expect a miracle here - strftime is an extremely complex function, it is OS-dependent AND locale-dependent, and we will surely find corner cases similar to the ones already found in pandas.

Do you have examples PRs merged in the past where the pandas team managed to replace an existing table computation function with a pyarrow compute function ? This could help me understand how you would like this to be done.

My recommendation still, would be to be very pragmatic here, we cannot do everything at once: a first step is to have an API to select engines, introducing a faster default engine with str.format. A second step will be to evaluate pyarrow strftime as an engine. I agree with you that this has potential to replace the OS strftime, but I'm afraid of the time and effort that it could take so I am reluctant to open the pandora box again and enter an 8 months-long contribution and review period without guarantee of success...

smarie commented:
In order to try to answer to your question about expected performance with pyarrow : I started to look at apache arrow internals and I could not yet find evidence that the arrow engine parses the string format specifier once upfront, into an efficient representation, to avoid re-parsing it for every cell of the TimestampArray. If you are more at ease with arrow codebase please do not hesitate to guide me here. I found this :https://github.com/search?q=repo%3Aapache%2Farrow%20TimestampFormatter&type=code
but this does not lead me anywhere yet

WillAyd commented:
@smarie if you are asking about where pyarrow implements this feature, it's in the C++ core:

https://github.com/apache/arrow/blob/c455d6b8c4ae2cb22baceb4c27e1325b973d39e1/cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc#L1196

The formatter is created once up front on L1199 before being applied element wise (see L1211-1216). In between there is also some heuristic for pre-allocating a string buffer (L1204-L1208), so my expectation is that it's performance will be hard to beat

> Do you have examples PRs merged in the past where the pandas team managed to replace an existing table computation function with a pyarrow compute function

pandas/core/arrays/string_arrow.py uses pyarrow.compute rather heavily, if pyarrow is installed. I think doing something similar here would be preferable 

> I'm afraid of the time and effort that it could take so I am reluctant to open the pandora box again and enter an 8 months-long contribution and review period without guarantee of success...

I understand and am really sympathetic to this. I don't think I've explicitly said this but THANK YOU for your time and effort on this.

The problem is though, that this works both ways. Larger PRs take a lot of time to write (as you've noticed) but also take a very long time to review (as you've seen from us), while posing a larger maintenance risk. Generally, I'd suggest getting alignment on the discussion and planning portion of a large, multi-file, API-changing initiative before diving in, as that makes the process more tenable for all parties involved

smarie commented:
Thanks @WillAyd . 

> I'd suggest getting alignment on the discussion and planning portion of a large, multi-file, API-changing initiative before diving in, as that makes the process more tenable for all parties involved

Agreed. Well, at first (as always ? :) ) it seemed like a straightforward, simple change to do. But as I discovered in the past 2 years implementing this (PR reviews started with @jbrockmendel and @mroeschke in 2022), many intermediate issues were hiding. I therefore had to solve first #46361, #46759, #47570, #46405, #53003, #51459 .

Unfortunately even now, this PR is still large and impacts a lot of files. Surprisingly, this is not due to the fact that the proposed engine is custom - indeed the proposed engine is just a single small file ([pandas/_libs/tslibs/strftime.py](https://github.com/pandas-dev/pandas/pull/51298/files)). It is due to the fact that pandas implementation of datetime and period formatting is still quite layered and with a few design inconsistencies. It is still much better than a couple months ago, as the team (@mroeschke I believe) simplified many useless alternate formatting functions. 

Here is a global picture of files impacted by this PR

![image](https://github.com/user-attachments/assets/68eeee9d-091c-4020-93a5-bfc3e7a4f806)

In addition, 9 test files, 3 asv benchmark files, and a couple init/api/meson files are updated.

As you can hopefully see if you use this picture to navigate the PR contents, adding the pyarrow engine would not change much of the complexity here - hence the PR would still be hard to review.

Based on this map, could we define together a target implementation strategy ? I am ready to break this PR into bits if you believe that this would make it easier to review.

smarie commented:
> Maybe answering that is the right approach forward - is there a particular design pattern we can apply that helps untangle that web?

I would work backwards from mandatory API up. So

 - We need scalar operations `Timestamp.strftime` and `Period.strftime` and they are there already - nothing to change there except introducing an `engine` parameter when people will want to force use a specific engine for a scalar (for debugging purposes or to get some engine-specific implementation of some strftime directive, as they are not all standard as you know)
 - We need array operations on classes `DatetimeArray`, `PeriodArray`, `DatetimeIndex`, `PeriodIndex`. `strftime` is everywhere provided by `DatetimeLikeArrayMixin.strftime`, which delegates to `cls._format_native_types`. Two particular entry points exist : `DatetimeArray._format_native_types` and `PeriodArray._format_native_types`. Here again everything seems fine to me, apart from adding the `engine` parameter

Now I see two things that are probably to improve: 

1. can we get rid of `get_format_datetime64`, `_format_datetime64`, and `_format_datetime64_dateonly` ? That would remove the bottom part of the picture. They seem to be used only in `DatetimeIndex._formatter_func` and in `_Datetime64TZFormatter._format_strings`. We could probably have `DatetimeIndex._formatter_func` leverage `DatetimeIndex.strftime` instead, and I'm pretty sure that `_Datetime64TZFormatter` does not need to override the `_format_strings` method inherited from `_Datetime64Formatter` anymore so we could just drop this overridance.

2. then, we could try to harmonize `period_array_strftime` and `format_array_from_datetime`. It seems that the design for periods is slightly better today than the one for timestamps: 

    - for **periods**, everything is in the same file `_libs/tslib/period.pyx`, and the core formatter today used both for arrays and scalars, is a function `cdef period_format` that leverages `libc.time.strftime` (os-dependent). Neither the `Period` class, nor the `period_array_strftime` function, contain any "shortcut" to perform formatting themselves, they both nicely delegate to `cdef period_format`. 
    - while for **timestamps** there is one entry point for scalars in `_libs/tslib/timestamps.pyx`, `Timestamp.strftime`, which delegates to `datetime.strftime` (stdlib). The latter is, at least in cpython implementation, using libc - but it is distribution-dependent. In parallel as opposed to what is done for periods, arrays are handled in `_libs/tslib.pyx` by `format_array_from_datetime`, that performs some formatting on its own in some cases, and in other cases delegates to `Timestamp.strftime`.

So for (2) I would recommend to move all of `_libs/tslib.pyx`'s content to `_libs/tslibs/timestamps.pyx` or to a dedicated `_libs/tslibs/datetimes.pyx`, and have the array (`format_array_from_datetime`) and scalar (`Timestamp.strftime`) methods both leverage the same (new) function `cdef timestamp_format` in there.

smarie commented:
@WillAyd any thoughts or feedback yet about the above plan ?

WillAyd commented:
> 1. can we get rid of `get_format_datetime64`, `_format_datetime64`, and `_format_datetime64_dateonly` ? That would remove the bottom part of the picture in the previous post. They seem to be used only in `DatetimeIndex._formatter_func` and in `_Datetime64TZFormatter._format_strings`. We could probably have `DatetimeIndex._formatter_func` leverage `DatetimeIndex.strftime` instead, and I'm pretty sure that `_Datetime64TZFormatter` does not need to override the `_format_strings` method inherited from `_Datetime64Formatter` anymore so we could just drop this overridance.

Those seem like good suggestions

> 2\. then, we could try to harmonize `period_array_strftime` and `format_array_from_datetime`. It seems that the design for periods is slightly better today than the one for timestamps:

This entire part seems to be in a tangle. So we have a mix of class methods and free-standing functions being responsible for the formatting? I'm not sure that going back to the class-method approach is the best; maybe there should just be a single class responsible for Strftime formatting the various objects to help consolidate the logic and follow a better separation of concerns

There's probably some history to the mixed design approaches that @MarcoGorelli and @jbrockmendel can offer guidance on

smarie commented:
Thanks @WillAyd . Could someone (@MarcoGorelli, @jbrockmendel ) confirm that I can open a (distinct) PR for 1. ?

Also could any of you refine or confirm item 2. ? I can definitely code any of the above proposals, but as advised by @WillAyd I would rather now make sure "in advance" that this has chances to match your expectations so as to be merged smoothly

thanks again !

MarcoGorelli commented:
> Could someone (@MarcoGorelli, @jbrockmendel ) confirm that I can open a (distinct) PR for 1. ?

sure, no objections from my side

auderson commented:
https://github.com/apache/arrow/blob/c455d6b8c4ae2cb22baceb4c27e1325b973d39e1/cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc#L1196
> 
> The formatter is created once up front on L1199 before being applied element wise (see L1211-1216). In between there is also some heuristic for pre-allocating a string buffer (L1204-L1208), so my expectation is that it's performance will be hard to beat

The formatter seems to be a wrapper around `arrow_vendored::date::to_stream`:
https://github.com/apache/arrow/blob/c3601a97a0718ae47726e6c134cbed4b98bd1a36/cpp/src/arrow/compute/kernels/temporal_internal.h#L146C7-L146C38

So I guess it's still called on each element. 
Here's the comparison as the way done in #44764

![image](https://github.com/user-attachments/assets/ab6bcdb2-99e2-4bc1-824d-d810a5970d6c)

Parvezkhan0 commented on asv_bench/benchmarks/strftime.py:
```diff
@@ -19,37 +19,68 @@ def setup(self, nobs):
                 "r": [np.random.uniform()] * nobs,
             }
         )
+        if tz_aware:
+            self.data["dt"] = self.data["dt"].dt.tz_localize("UTC")
+            self.data["d"] = self.data["d"].dt.tz_localize("UTC")
+
+        self.data["i"] = self.data["dt"]
+        self.data.set_index("i", inplace=True)
```
The parameter tz_aware is being used to toggle between timezone-aware and naive timestamps. However, the logic for handling timezone-aware data in the setup method is somewhat repetitive and could benefit from encapsulation in a utility function.

Extract the logic for timezone localization into a helper function to improve readability and maintainability.
For example:
`
def localize_if_required(dataframe, tz_aware):
    if tz_aware:
        dataframe["dt"] = dataframe["dt"].dt.tz_localize("UTC")
        dataframe["d"] = dataframe["d"].dt.tz_localize("UTC")

`

Parvezkhan0 reviewed (changes requested):
Update the documentation to include an example of when the fallback to OS-level strftime occurs and its performance implications.

smarie commented:
Thanks @auderson and @Parvezkhan0 , this will provide some motivation fuel for taking a new stab at this :) most probably in the quieter period after end of year

mroeschke commented:
Thanks for the effort in this PR over the years. Unfortunately, it appears to have lost steam and probably needs more direction from a core team member to properly solve the OP. Since the PR has gone stale in the meantime going to close
